### PR TITLE
User authentication

### PR DESCRIPTION
**Pull request by leechongyan**
Amended code structure with chin changes

### CODE REVIEW

**Comment by seaerchin** [comment]
btw i think you shouldn't merge before approval

**Comment by leechongyan** [comment]
yup i merged cos the main has the older version of mine, would like to update it, u can still comment and thats why this branch is still open

File: authentication_service/controllers/account_handlers.go [hunk 1]
```diff
@@ -61,7 +61,11 @@ func SignUp() gin.HandlerFunc {
 		database.UserCollection[*user.Email] = user
 
 		// send an email
-		mail_service.SendVerificationCode(user, new_V_key)
+		err = mail_service.SendVerificationCode(user, new_V_key)
+		if err != nil {
+			c.JSON(http.StatusOK, gin.H{"error": err.Error()})
```
Comment (seaerchin) [inline authentication_service/controllers/account_handlers.go:66]: idt we should return 200 if there's an error ._.

**Review by seaerchin — COMMENTED** [review]
1. we should consider returning custom error types so that we can use a type switch and return more information back to the user/let the handler do more work
2. we should write a wrapper over the db. this is because what happens here is our abstraction is leaky - our upper layers has knowledge of the concrete type of the database. this breaks when we wish to change database, especially because we are going to use sql for this
3. why are we wrapping all our routes in a higher order function, which is then invoked without any arguments? this makes the code harder to read + hides intent, which is why i'd suggest removing it!

will review more thoroughly tmrw

**Pull request merged**